// lib/errors.js
'use strict';

const kPacket = Symbol('packet');
const kSender = Symbol('sender');

/**
 * Base class for STUN errors.
 */
class StunError extends Error {
  /**
   * @class StunError
   * @param {string} message Error message.
   * @param {Buffer|StunMessage} packet Arrived data.
   * @param {Object} sender
   */
  constructor(message, packet, sender) {
    super(message);

    this.name = 'StunError';
    this[kPacket] = packet;
    this[kSender] = sender;
  }

  /**
   * Get the pinned packet.
   * @returns {Buffer|StunMessage|undefined}
   */
  get packet() {
    return this[kPacket];
  }

  /**
   * Get remote info.
   * @returns {Object|undefined}
   */
  get sender() {
    return this[kSender];
  }
}

/**
 * Error class for invalid arrived messages.
 */
class StunMessageError extends StunError {
  /**
   * @class StunInvalidMessage
   * @param {Buffer|StunMessage} packet Arrived data.
   * @param {Object} sender
   */
  constructor(packet, sender) {
    super('Invalid message', packet, sender);

    this.name = 'StunMessageError';
  }
}

module.exports = {
  StunError,
  StunMessageError,
};
