// net/request.js
'use strict';

const parseUrl = require('parse-url');
const debug = require('debug')('stun:request');
const StunServer = require('net/dgram-server');
const StunRequest = require('message/request');
const { messageType, eventNames } = require('lib/constants');
const { createServer } = require('net/create-server');
const { createMessage } = require('lib/create-message');

module.exports = {
  request,
};

/**
 * @typedef {Object} RequestOptions
 * @property {StunServer} [server]
 * @property {dgram.Socket} [socket]
 * @property {StunRequest} [message]
 * @property {number} [timeout] Initial RTO.
 * @property {number} [maxTimeout] Maximal RTO.
 * @property {number} [retries] Maximal the number of retries.
 */

const defaultRetryOptions = {
  retries: 6,
  timeout: 500,
  maxTimeout: Infinity,
};

/**
 * Send a stun binding request to resolve ip address.
 * @param {string} url
 * @param {RequestOptions} [options]
 * @param {Function} callback
 */
function request(url, options = {}, callback) {
  if (typeof options === 'function') {
    callback = options; // eslint-disable-line no-param-reassign
    options = {}; // eslint-disable-line no-param-reassign
  }

  const { port, protocol, resource } = parseUrl(url, {
    defaultProtocol: 'stun://',
    stripWWW: false,
  });

  if (protocol !== 'stun') {
    process.nextTick(callback, new Error(`Invalid protocol '${protocol}'`));
    return;
  }

  debug('start request to %s:%s', resource, port || 3478);

  let server;
  let message;
  let externalServer = false;

  if (options.server instanceof StunServer) {
    debug('use external server');
    server = options.server; // eslint-disable-line prefer-destructuring
    externalServer = true;
  } else {
    debug('create server');
    const { socket } = options;
    server = createServer({ type: 'udp4', socket });
  }

  if (options.message instanceof StunRequest) {
    debug('use external message');
    message = options.message; // eslint-disable-line prefer-destructuring
  } else {
    debug('create BINDING_REQUEST message');
    message = createMessage(messageType.BINDING_REQUEST);
  }

  const retrier = () => {
    debug('send message');
    server.send(message, port || 3478, resource);
  };

  const retryOptions = Object.assign({}, defaultRetryOptions, options);
  const done = retry(retrier, retryOptions, (err, res) => {
    cleanup();
    process.nextTick(callback, err, res);
  });

  server.on(eventNames.EVENT_BINDING_RESPONSE, onResponse);
  server.on('error', onError);

  /**
   * Error handler.
   * @param {Error} err
   */
  function onError(err) {
    // check stun err for transation

    debug('got stun error %s', err.message);
    done(err);
  }

  /**
   * Handle incoming binding_response messages.
   * @param {StunResponse} res
   */
  function onResponse(res) {
    debug('got response');
    if (Buffer.compare(message.transactionId, res.transactionId) !== 0) {
      debug('invalid transaction id, skip');
      return;
    }

    debug('success');
    done(null, res);
  }

  /**
   * Remove request's handlers.
   */
  function cleanup() {
    debug('clean up');
    server.removeListener('error', onError);
    server.removeListener(eventNames.EVENT_BINDING_RESPONSE, onResponse);

    if (!externalServer) {
      debug('close server');
      server.close();
    }
  }
}

/**
 *
 * @param {Function} fn
 * @param {Object} opts
 * @param {number} opts.timeout Initial RTO.
 * @param {number} opts.maxTimeout Maximal RTO.
 * @param {number} opts.retries Maximal the number of retries.
 * @param {Function} callback
 * @returns {Function}
 */
function retry(fn, opts, callback) {
  let handle;
  debug('retry options %o', opts);

  /**
   * Clean up.
   * @param {Error} err
   * @param {Object} res
   */
  function done(err, res) {
    if (handle) {
      clearTimeout(handle);
    }

    debug('done');
    callback(err, res);
  }

  /**
   * Runner.
   * @param {number} count
   */
  function attempt(count) {
    const next = () => {
      debug('attempt %s', count);
      fn();
      attempt(count + 1);
    };

    if (count === 0) {
      process.nextTick(next);
      return;
    }

    const retries = count - 1;
    const rto = opts.timeout << retries; // eslint-disable-line no-bitwise
    const timeout = Math.min(rto, opts.maxTimeout);
    debug('start timer %s (rto = %s)', timeout, rto);

    if (count > opts.retries) {
      debug('max retries %s / %s', retries, opts.retries);
      handle = setTimeout(() => done(new Error('timeout')), timeout);
    } else {
      handle = setTimeout(next, timeout);
    }
  }

  attempt(0);
  return done;
}
